### DISCUSSION TITLE
Hirercial Reinforcement Learning

### DISCUSSION THREAD

**Discussion by PhanindraParashar** [eleurent/highway-env#224] [Ideas]
I was loking at some hirercial methods like option critic, PPOC and others. What I think is that if we divide the task into several hirercies we could fasten the learning process.

1st level learns to pick a lane
2nd level learns to pick a waypoint to reach
3rd level (controller) would learn to execute

Any similar ideas? 
Has anyone tried this before?

**Comment by eleurent** [comment]
I don't recall having seen "pure" hierarchical RL applied in the context of autonomous driving (though I haven't looked for it specifically).
Yet, this kind of architecture (pick lane -> pick waypoint -> low-level control) is very common in traditional robotics pipelines (i.e. handcrafted, non-learning-based solutions) and self-driving software, so I think that it makes a lot of sense.
And there are several papers where the policy is separated into high-level vs low-level parts, e.g. https://iliad.stanford.edu/pdfs/publications/cao2020reinforcement.pdf and https://openreview.net/pdf?id=Skl4mRNYDr

But I haven't tried anything like that myself, so if you go for it I'd be curious to see your results.

**Reply by PhanindraParashar** [reply]
I have seen a paper :

Trajectory Planning for Autonomous Vehicles Using Hierarchical Reinforcement Learning

Link: https://arxiv.org/abs/2011.04752

Where hirercial reinforcement learning is used for the autonomous driving. 

I am currently working on the similar ideas from the robotics perspective where we do the fallowing tasks 

Pick lane >> pick waypoint and generate trajectory >> use low level control like PID to fallow trajectory 

I am currently working on some reward functions that could be used to pick a lane and a saperate reward function for how good the trajectory is. 

I was loking on ways to constrain the network to generate a waypoint that is feasible (according to vehicle constraints on max speed and max acceleration). 

I am currently thinking of using polynomial (5th order) for smooth trajectory generation and using autograd to find peak acceleration (by differentiation of the polynomial) and manually getting constraints for the maximum and minimum coordinates of waypoints. And use this as constraint to output of the network that gives waypoints. Is this approach correct? 

Are there any ideas?
